### ISSUE TITLE
CS0122 wrong member inaccessible when overload does not exist

### ISSUE THREAD

**Issue by miloush** [dotnet/roslyn#15401]
**Version Used**:  2.0.0.61104

**Steps to Reproduce**:

.NET < 4.6

```C#
using System.Net.Sockets;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            TcpClient tcp = new TcpClient();
            tcp.Dispose();
        }
    }
}
```

**Expected Behavior**: `CS1061 'TcpClient' does not contain a definition for 'Dispose' ...`

**Actual Behavior**: `CS0122 'TcpClient.Dispose(bool)' is inaccessible due to its protection level`

This happens for example when downgrading project from 4.6 (which has TcpClient.Dispose method) to 4.5.

**Comment by svick** [comment]
Could you explain why do you think that error would be better? Especially considering that the error message would be technically wrong (since `TcpClient` does contain a definition for `Dispose`).

**Comment by miloush** [comment]
@svick 

> since  TcpClient  does contain a definition for  Dispose 

No it does not in 4.5, that is the point (and reason for the error in the first place).

The actual error CS0122 is very confusing because it tells you that `Dispose(bool)` is inaccessible but you are not calling `Disposible(bool)` anywhere. So it's like _.... well, that's true I guess.. so what?_

**Comment by svick** [comment]
The way I see it, the compiler does this:

1. Find out if there are any methods called `Dispose`.
2. Find out if any of them are accessible.
3. Find out which one to call, based on overload resolution rules.

In this case, step 1 succeeds (it finds `protected void Dispose(bool)`), step 2 fails (because it's `protected`) and so that is the error that is reported.

Step 3 is never reached, so the compiler does not get a chance to figure out that `protected void Dispose(bool)` is not applicable also because it expects a parameter.

Though just because it follows the logic of the compiler (or the specification) does not necessarily mean it's a good error message. But I think in this case it is, it's better if the compiler tells you why the `Dispose` method it found couldn't be used, than if it pretended there is no `Dispose` method.

Put another way, `protected void Dispose(bool)` is wrong for two reasons (accessibility and expecting parameter), and I would expect the compiler to report one of them, which is what it does.

**Comment by miloush** [comment]
Hmm I see your point. You are also implying that the overload resolution happens only on accessible methods, I guess for performance reasons. The fact that it does not report the called overload even if it does exist makes it even more confusing:
```C#
public class Test
{
     private void Method(bool c) {}
     private void Method(int a, int b) { }
}
...
t.Method(2, 4); // Test(bool) is inaccessible
```
Maybe the error could somehow reflect the facts you described?

**Comment by miloush** [comment]
The same way you could argue that `No overload for method 'Method' takes 2 arguments.` (should the first one be public) is technically wrong and it should be `No *accessible* overload for method 'Method' takes 2 arguments.` instead.

**Comment by svick** [comment]
> Maybe the error could somehow reflect the facts you described?

Yeah, if that error message actually means "there is no accessible overload", as I think it does, then I would prefer if it didn't specify an overload. I.e. change:

> CS0122 'Test.Method(bool)' is inaccessible due to its protection level

To:

> CS0122 'Test.Method' is inaccessible due to its protection level

**Comment by CyrusNajmabadi** [comment]
An improved error message here would be desirable.  However, it's still important that hte compiler *under the covers* bind to these members and expose them through the API.

THis is, for example, how Roslyn can ensure that things like goto-def work across langauge from VB to C# (and vice versa).  If you have a VB class where you define ```private Dispose(bool)``` and you then reference that from C# as "Dispose", we want to get the message that you messed things up, and we want to be able to allow you to goto-def to get to that symbol, as well as provide suitable 'fixes' (in the future at least) for things like changing the accessibility and whatnot.
